assembly/src/parsers/u32_ops.rs

```diff
@@ -182,81 +252,140 @@ pub fn parse_u32split(span_ops: &mut Vec<Operation>, op: &Token) -> Result<(), A
-    match op.num_parts() {
-        0 => return Err(AssemblyError::missing_param(op)),
-        1 => span_ops.push(Operation::U32sub),
-        _ => return Err(AssemblyError::extra_param(op)),
-    }
-
-    Ok(())
+    handle_arthimetic_operation(span_ops, op, Operation::U32sub, true)
 }
 
 /// Translates u32mul assembly instruction to VM operation U32MUL.
+/// Depending on the different mode, additional instructions will be inserted.
+/// Please refer to the docs of `handle_arthimetic_operation` for more details.
+/// VM cycles per mode:
+/// - u32mul: 10 cycles
+/// - u32mul.b: 7 cycles
+/// - u32mul.full: 8 cycles
+/// - u32mul.unsafe: 1 cycle
 pub fn parse_u32mul(span_ops: &mut Vec<Operation>, op: &Token) -> Result<(), AssemblyError> {
-    match op.num_parts() {
-        0 => return Err(AssemblyError::missing_param(op)),
-        1 => span_ops.push(Operation::U32mul),
-        _ => return Err(AssemblyError::extra_param(op)),
-    }
-
-    Ok(())
+    handle_arthimetic_operation(span_ops, op, Operation::U32mul, false)
 }
 
 /// Translates u32madd assembly instruction to VM operation U32MADD.
+/// In safe mode, we assert all three nummbers are u32.
+/// VM cycles per mode:
+/// - u32madd: 12 cycles
+/// - u32madd.unsafe: 1 cycle
 pub fn parse_u32madd(span_ops: &mut Vec<Operation>, op: &Token) -> Result<(), AssemblyError> {
     match op.num_parts() {
         0 => return Err(AssemblyError::missing_param(op)),
-        1 => span_ops.push(Operation::U32madd),
+        1 => {
+            assert_u32(span_ops);
+            span_ops.push(Operation::Swap);
+            assert_u32(span_ops);
+            span_ops.push(Operation::MovUp2);
+            assert_u32(span_ops);
+        }
+        2 => {
+            if op.parts()[1] != "unsafe" {
+                return Err(AssemblyError::invalid_param(op, 1));
+            }
+        }
         _ => return Err(AssemblyError::extra_param(op)),
     }
 
+    span_ops.push(Operation::U32madd);
+
     Ok(())
 }
 
-/// Translates u32div assembly instruction to VM operation U32DIV.
+/// Translates u32addc assembly instruction to VM operation U32DIV.
+/// Depending on the different mode, additional instructions will be inserted.
+/// Please refer to the docs of `handle_arthimetic_operation` for more details.
+/// VM cycles per mode:
+/// - u32div: 11 cycles
+/// - u32div.b: 7 cycles
+/// - u32div.full: 9 cycles
+/// - u32div.unsafe: 1 cycle
 pub fn parse_u32div(span_ops: &mut Vec<Operation>, op: &Token) -> Result<(), AssemblyError> {
-    match op.num_parts() {
-        0 => return Err(AssemblyError::missing_param(op)),
-        1 => span_ops.push(Operation::U32div),
-        _ => return Err(AssemblyError::extra_param(op)),
-    }
-
-    Ok(())
+    handle_arthimetic_operation(span_ops, op, Operation::U32div, true)
 }
 
-/// Translates u32mod assembly instruction to VM operation U32DIV DROP.
+/// Translates u32mod assembly instruction to VM operation U32SPLIT DROP.
```

Comment:
What we actually need to do in this operation is `U32DIV SWAP DROP`. We are computing `a % b` (where `b` is at the top of the stack). Doing `U32SPLIT DROP` just does `b % 2^32`. But doing `U32DIV SWAP DROP` keeps the remainder on the top of the stack.